File: TypeScript/src/drawing/shapes/shape.ts
import { Writer } from "../writing/writer";
import { PNGWriter } from "../writing/pngwriter";
import { JPEGWriter } from "../writing/jpegwriter";
import { Line } from "./line";

/**
 * Refactor Exercise 3: (Mis-)Shaped
 *
 * @author Zishen Wen (F22), Deyuan Chen (S22)
 */
interface Shape {
	
    toLines(): Line[];

    draw(writer: Writer, lines: Line[]): void;
}

abstract class AbstractShape implements Shape {

	/**
     * Converts shape into lines.
     *
     * @return lines of this shape.
     */
	abstract toLines(): Line[];

	/**
     * Draws lines to file.
     */
	public draw(writer: Writer, lines: Line[]): void {
		try {
			for (let line of lines) {
				// TODO: what is the purpose of the code there?
				if (writer instanceof JPEGWriter) {
					writer.write(line.toJPEG());
				}
				else if (writer instanceof PNGWriter) {
					writer.write(line.toPNG());
				}
			}
		} catch (e) {
			console.log(e);
		}
	}
}

export { Shape, AbstractShape }

File: Java/src/main/java/drawing/shapes/Shape.java
package drawing.shapes;

import drawing.writing.JPEGWriter;
import drawing.writing.PNGWriter;

import java.io.IOException;
import java.io.Writer;

/**
 * Refactor Task 3: (Mis-)Shaped
 *
 * @author Zishen Wen (F22), Deyuan Chen (S22)
 */
public interface Shape {

    /**
     * Converts shape into lines.
     *
     * @return lines of this shape.
     */
    Line[] toLines();

    /**
     * Draws lines to file.
     */
    default void draw(Writer writer, Line[] lines) {
        try {
            for (Line line : lines) {
                // TODO: what is the purpose of the code there?
                if (writer instanceof JPEGWriter) {
                    writer.write(line.toJPEG());
                } else if (writer instanceof PNGWriter) {
                    writer.write(line.toPNG());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
